File: src/main/java/komu/blunt/types/checker/Assumptions.java
package komu.blunt.types.checker;

import komu.blunt.eval.TypeCheckException;
import komu.blunt.objects.Symbol;
import komu.blunt.types.Scheme;
import komu.blunt.types.TypeVariable;
import komu.blunt.types.Types;

import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Set;

import static com.google.common.base.Preconditions.checkArgument;
import static com.google.common.base.Preconditions.checkNotNull;
import static java.util.Collections.emptyMap;
import static java.util.Collections.unmodifiableMap;

public final class Assumptions implements Types<Assumptions> {

    private final Map<Symbol, Scheme> mappings;
    
    private Assumptions() {
        this.mappings = emptyMap();
    }

    private Assumptions(Builder builder) {
        this.mappings = unmodifiableMap(builder.mappings);
    }

    public Scheme find(Symbol name) {
        Scheme scheme = mappings.get(name);
        if (scheme != null)
            return scheme;
        else
            throw new TypeCheckException("unbound identifier: " + name);
    }

    @Override
    public String toString() {
        return mappings.toString();
    }

    @Override
    public void addTypeVariables(Set<TypeVariable> variables) {
        for (Scheme scheme : mappings.values())
            scheme.addTypeVariables(variables);
    }

    @Override
    public Assumptions apply(Substitution substitution) {
        Builder builder = builder();
     
        for (Map.Entry<Symbol, Scheme> entry : mappings.entrySet())
            builder.add(entry.getKey(), entry.getValue().apply(substitution));

        return builder.build();
    }

    public static Assumptions singleton(Symbol arg, Scheme scheme) {
        return builder().add(arg, scheme).build();
    }

    public Assumptions join(Assumptions as) {
        return builder().addAll(as).addAll(this).build();
    }

    public static Assumptions from(List<Symbol> names, List<Scheme> schemes) {
        checkArgument(names.size() == schemes.size(), names.size() + " != " + schemes.size());

        Builder builder = new Builder();
        for (int i = 0; i < names.size(); i++)
            builder.add(names.get(i), schemes.get(i));
        
        return builder.build();
    }
    
    public static Builder builder() {
        return new Builder();
    }

    public static Assumptions empty() {
        return new Assumptions();
    }

    public static final class Builder {
        private Map<Symbol, Scheme> mappings = new HashMap<>();
        private boolean built = false;

        public Builder add(Symbol name, Scheme scheme) {
            ensurePrivateCopy();
            mappings.put(checkNotNull(name), checkNotNull(scheme));
            return this;
        }

        public Builder addAll(Assumptions as) {
            ensurePrivateCopy();
            mappings.putAll(as.mappings);
            return this;
        }
        
        public Assumptions build() {
            built = true;
            return new Assumptions(this);
        }

        public Assumptions build(Assumptions as) {
            return build().join(as);
        }
        
        private void ensurePrivateCopy() {
            if (built) {
                mappings = new HashMap<>(mappings);
                built = false;
            }
        }
    }
}

File: src/main/kotlin/komu/blunt/types/checker/Assumptions.kt
package komu.blunt.types.checker

import komu.blunt.eval.TypeCheckException
import komu.blunt.objects.Symbol
import komu.blunt.types.Scheme
import komu.blunt.types.Type
import komu.blunt.types.Types
import java.util.*
import java.util.Collections.emptyMap
import java.util.Collections.singletonMap

class Assumptions (private val mappings: Map<Symbol,Scheme>) : Types<Assumptions> {

    fun join(ass: Assumptions) =
        builder().addAll(ass).addAll(this).build()

    fun find(name: Symbol): Scheme =
        mappings[name] ?: throw TypeCheckException("unbound identifier: '$name'")

    override fun toString() = mappings.toString()

    override fun addTypeVariables(result: MutableSet<Type.Var>) {
        for (scheme in mappings.values)
            scheme.addTypeVariables(result)
    }

    override fun apply(substitution: Substitution): Assumptions {
        val builder = builder()

        for ((key, value) in mappings)
            builder[key] = value.apply(substitution)

        return builder.build()
    }

    companion object {

        fun builder() = Builder()
        fun empty() = Assumptions(emptyMap())
        fun singleton(arg: Symbol, scheme: Scheme) = Assumptions(singletonMap(arg, scheme))

        fun from(names: List<Symbol>, schemes: List<Scheme>): Assumptions {
            if (names.size != schemes.size)
                throw IllegalArgumentException("${names.size} != ${schemes.size}")

            val builder = Builder()
            for (i in names.indices)
                builder[names[i]] = schemes[i]

            return builder.build()
        }

        class Builder {

            private var mappings = HashMap<Symbol,Scheme>()
            private var built = false

            operator fun set(name: Symbol, scheme: Scheme) {
                ensurePrivateCopy()
                mappings[name] = scheme
            }

            fun addAll(ass: Assumptions): Builder {
                ensurePrivateCopy()
                mappings.putAll(ass.mappings)
                return this
            }

            fun build(): Assumptions {
                built = true
                return Assumptions(mappings)
            }

            fun build(ass: Assumptions) = build().join(ass)

            private fun ensurePrivateCopy() {
                if (built) {
                    mappings = HashMap(mappings)
                    built = false
                }
            }
        }
    }
}
